x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts

```diff
@@ -447,12 +447,94 @@ export const privilege = t.type({
 
 export type Privilege = t.TypeOf<typeof privilege>;
 
-export enum BulkAction {
+export enum BulkActionSimple {
   'enable' = 'enable',
   'disable' = 'disable',
   'export' = 'export',
   'delete' = 'delete',
   'duplicate' = 'duplicate',
 }
 
+export const bulkActionSimple = enumeration('BulkActionSimple', BulkActionSimple);
+
+export enum BulkActionEnhanced {
+  'update' = 'update',
+}
+export const bulkActionEnhanced = enumeration('BulkActionEnhanced', BulkActionEnhanced);
+
+export const BulkAction = {
+  ...BulkActionSimple,
+  ...BulkActionEnhanced,
+};
+
+export type BulkAction = BulkActionSimple | BulkActionEnhanced;
```

Comment:
Can we avoid splitting bulk action into multiple enums: `BulkActionSimple` and `BulkActionEnhanced`? I think the current approach is not easily extensible. For example, if we would need to add more actions, we had to add one more enum, as `BulkActionEnhanced` is designed to hold exactly one action: `update`, and adding more would break this validation:

```ts
t.type({
  query: queryOrUndefined,
  action: bulkActionEnhanced,
  updates: t.array(bulkActionUpdate),
})
```

Reply:
I did not find a way to do this and keep validation correct.

I split it into 2 enums as we essentially have 2 types of actions:
1. simpler actions, with just 2 properties: action and query
2. update(enhanced) with 3 properties: action, query, updates.

So, in this case: `updates` property will be required for `update` action only, and will not be validated for other actions.

Initially, I did not split actions into 2 enums, but put updates property as partial. But downside in that approach was I had to do additional validation when processing actions in route function. And essentially that solution would not take full advantage of request's body validation we have.

> For example, if we would need to add more actions, we had to add one more enum, as BulkActionEnhanced is designed to hold exactly one action: update, and adding more would break this validation:

I think it would depend on type of action we want to introduce: if it does not require any additional properties it can be put into `BulkActionSimple`. If it would require another unique property, specific to this action, we would need to create a separate schema for this action to validate it.

Reply:
Just tried to do that with simple unions, validation is working fine:

```diff
diff --git a/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts b/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts
index 47e21487bf7..3ba26db1468 100644
--- a/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts
+++ b/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts
@@ -447,27 +447,14 @@ export const privilege = t.type({
 
 export type Privilege = t.TypeOf<typeof privilege>;
 
-export enum BulkActionSimple {
+export enum BulkAction {
   'enable' = 'enable',
   'disable' = 'disable',
   'export' = 'export',
   'delete' = 'delete',
   'duplicate' = 'duplicate',
-}
-
-export const bulkActionSimple = enumeration('BulkActionSimple', BulkActionSimple);
-
-export enum BulkActionEnhanced {
   'update' = 'update',
 }
-export const bulkActionEnhanced = enumeration('BulkActionEnhanced', BulkActionEnhanced);
-
-export const BulkAction = {
-  ...BulkActionSimple,
-  ...BulkActionEnhanced,
-};
-
-export type BulkAction = BulkActionSimple | BulkActionEnhanced;
 
 export const bulkAction = enumeration('BulkAction', BulkAction);
 
diff --git a/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts b/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts
index 4855087de4c..e4fcabd7868 100644
--- a/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts
+++ b/x-pack/plugins/security_solution/common/detection_engine/schemas/request/perform_bulk_action_schema.ts
@@ -6,24 +6,25 @@
  */
 
 import * as t from 'io-ts';
-import {
-  bulkActionEnhanced,
-  bulkActionSimple,
-  queryOrUndefined,
-  bulkActionUpdate,
-} from '../common/schemas';
+import { BulkAction, queryOrUndefined, bulkActionUpdate } from '../common/schemas';
 
 export const performBulkActionSchema = t.union([
   t.exact(
     t.type({
       query: queryOrUndefined,
-      action: bulkActionSimple,
+      action: t.union([
+        t.literal(BulkAction.delete),
+        t.literal(BulkAction.disable),
+        t.literal(BulkAction.duplicate),
+        t.literal(BulkAction.enable),
+        t.literal(BulkAction.export),
+      ]),
     })
   ),
   t.exact(
     t.type({
       query: queryOrUndefined,
-      action: bulkActionEnhanced,
+      action: t.literal(BulkAction.update),
       updates: t.array(bulkActionUpdate),
     })
   ),
```

Reply:
updated, thank for suggestion